pep-0695.rst

```diff
@@ -401,11 +420,9 @@ defined in an outer scope.
     # eliminate this limitation.
     class ClassA[S, T: Sequence[S]]: ...
 
-    # The following generates a compiler error or runtime exception because T
-    # is referenced before it is defined. This occurs even though T is defined
-    # in the outer scope.
-    T = 0
-    class ClassB[S: Sequence[T], T]: ...  # Compiler error: T is not defined
+    # The following generates no compiler error, because the bound for ``S``
```

Comment:
Pyright's current implementation enforces that the expression doesn't reference a type parameter later in the list. With this change, this will be legal at runtime, although it should still be a type-checker error (as you note below). I think that's the right answer, as I don't see a use case for it. It can also lead to perverse errors like circular dependencies.

```python
def class[S: list[T], T: list[S]]: ...
```

Reply:
Yes, I think this behavior is not useful, but I don't see a good way to disallow it while still using lazy evaluation and without greatly complicating the implementation.

Reply:
Bounds cannot be generic; they need to be concrete, so this wouldn't work. 

Default type argument types (a'la PEP 696) can refer to other type parameters (e.g. `class C[S, T=S]: ...`), but I can't think of a case where it would make sense to refer to a later type parameter in a default expression. For example, `class C[S=T, T]: ...` doesn't make sense.